## Лабораторна робота №4 
виконала студентка групи ІС-12 мп

Кравчук Олена

Варіант-9

In [0]:
import pyspark.sql.functions as F
from pyspark.sql.window import *

Створюємо схему та зчитуємо дані з файлу

In [0]:
schema = "`ID` INT,`Length` INT,`Birthweight` DOUBLE,`Headcirc` INT,`Gestation` INT,`smoker` INT,`mage` INT,`mnocig` INT,`mheight` INT,`mppwt` INT,`fage` INT,`fedyrs` INT,`fnocig` INT,`fheight` INT,`lowbwt` INT,`mage35` INT"

df = spark.read.format('csv').option('header', True).option('schema', schema).load("/FileStore/tables/Birthweight__1_.csv")
display(df)

ID Length Birthweight Headcirc Gestation smoker mage mnocig mheight mppwt fage fedyrs fnocig fheight lowbwt mage35 1360 56 4.55 34 44 0 20 0 162 57 23 10 35 179 0 0 1016 53 4.32 36 40 0 19 0 171 62 19 12 0 183 0 0 462 58 4.1 39 41 0 35 0 172 58 31 16 25 185 0 1 1187 53 4.07 38 44 0 20 0 174 68 26 14 25 189 0 0 553 54 3.94 37 42 0 24 0 175 66 30 12 0 184 0 0 1636 51 3.93 38 38 0 29 0 165 61 31 16 0 180 0 0 820 52 3.77 34 40 0 24 0 157 50 31 16 0 173 0 0 1191 53 3.65 33 42 0 21 0 165 61 21 10 25 185 0 0 1081 54 3.63 38 38 0 18 0 172 50 20 12 7 172 0 0 822 50 3.42 35 38 0 20 0 157 48 22 14 0 179 0 0 1683 53 3.35 33 41 0 27 0 164 62 37 14 0 170 0 0 1088 51 3.27 36 40 0 24 0 168 53 29 16 0 181 0 0 1107 52 3.23 36 38 0 31 0 164 57 35 16 0 183 0 0 755 53 3.2 33 41 0 21 0 155 55 25 14 25 183 0 0 1058 53 3.15 34 40 0 29 0 167 60 30 16 25 182 0 0 321 48 3.11 33 37 0 28 0 158 54 39 10 0 171 0 0 697 48 3.03 35 39 0 27 0 162 62 27 14 0 178 0 0 808 48 2.92 33 34 0 26 0 167 64 25 12 25 175 0 0 1600 53 2.9 34 39 0 19 0 165 57 23 14 2 193 0 0 1313 43 2.65 32 33 0 24 0 149 45 26 16 0 169 1 0 792 53 3.64 38 40 1 20 2 170 59 24 12 12 185 0 0 1388 51 3.14 33 41 1 22 7 160 53 24 16 12 176 0 0 575 50 2.78 30 37 1 19 7 165 60 20 14 0 183 0 0 569 50 2.51 35 39 1 22 7 159 52 23 14 25 200 1 0 1363 48 2.37 30 37 1 20 7 163 47 20 10 35 185 1 0 300 46 2.05 32 35 1 41 7 166 57 37 14 25 173 1 1 431 48 1.92 30 33 1 20 7 161 50 20 10 35 180 1 0 1764 58 4.57 39 41 1 32 12 173 70 38 14 25 180 0 0 532 53 3.59 34 40 1 31 12 163 49 41 12 50 191 0 0 752 49 3.32 36 40 1 27 12 152 48 37 12 25 170 0 0 1023 52 3 35 38 1 30 12 165 64 38 14 50 180 0 0 57 51 3.32 38 39 1 23 17 157 48 32 12 25 169 0 0 1522 50 2.74 33 39 1 21 17 156 53 24 12 7 179 0 0 223 50 3.87 33 45 1 28 25 163 54 30 16 0 183 0 0 272 52 3.86 36 39 1 30 25 170 78 40 16 50 178 0 0 27 53 3.55 37 41 1 37 25 161 66 46 16 0 175 0 1 365 52 3.53 37 40 1 26 25 170 62 30 10 25 181 0 0 619 52 3.41 33 39 1 23 25 181 69 23 16 2 181 0 0 1369 49 3.18 34 38 1 31 25 162 57 32 16 50 194 0 0 1262 53 3.19 34 41 1 27 35 163 51 31 16 25 185 0 0 516 47 2.66 33 35 1 20 35 170 57 23 12 50 186 1 0 1272 53 2.75 32 40 1 37 50 168 61 31 16 0 173 0 1

1) Створити нову таблицю, яка буде містити наступні два стовпці: вік
матері; масив ІD записів з дітьми матерів відповідного віку.

In [0]:
df1 = df.groupBy("mage") \
        .agg(F.collect_list("ID") \
        .alias("ID")) \
        .withColumnRenamed("mage", "Вік") 

display(df1)

Вік ID 29 List(1636, 1058) 30 List(1023, 272) 28 List(321, 223) 22 List(1388, 569) 35 List(462) 31 List(1107, 532, 1369) 18 List(1081) 27 List(1683, 697, 752, 1262) 26 List(808, 365) 19 List(1016, 1600, 575) 41 List(300) 23 List(57, 619) 24 List(553, 820, 1088, 1313) 32 List(1764) 20 List(1360, 1187, 822, 792, 1363, 431, 516) 37 List(27, 1272) 21 List(1191, 755, 1522)

2) Взяти таблицю з завдання 1 і перетворити другий стовпець на кілька
стовпців.

In [0]:
df2 =df1 \
.select('Вік', \
        F.explode('ID').alias('id'), \
        F.row_number().over(Window.partitionBy("Вік").orderBy("Вік")).alias('number'))

display(df2)

Вік id number 18 1081 1 19 1016 1 19 1600 2 19 575 3 20 1360 1 20 1187 2 20 822 3 20 792 4 20 1363 5 20 431 6 20 516 7 21 1191 1 21 755 2 21 1522 3 22 1388 1 22 569 2 23 57 1 23 619 2 24 553 1 24 820 2 24 1088 3 24 1313 4 26 808 1 26 365 2 27 1683 1 27 697 2 27 752 3 27 1262 4 28 321 1 28 223 2 29 1636 1 29 1058 2 30 1023 1 30 272 2 31 1107 1 31 532 2 31 1369 3 32 1764 1 35 462 1 37 27 1 37 1272 2 41 300 1

In [0]:
df2 = df2 \
    .groupby('Вік') \
    .pivot('number') \
    .agg(F.first('id'))

display(df2)

Вік 1 2 3 4 5 6 7 18 1081 null null null null null null 19 1016 1600 575 null null null null 20 1360 1187 822 792 1363 431 516 21 1191 755 1522 null null null null 22 1388 569 null null null null null 23 57 619 null null null null null 24 553 820 1088 1313 null null null 26 808 365 null null null null null 27 1683 697 752 1262 null null null 28 321 223 null null null null null 29 1636 1058 null null null null null 30 1023 272 null null null null null 31 1107 532 1369 null null null null 32 1764 null null null null null null 35 462 null null null null null null 37 27 1272 null null null null null 41 300 null null null null null null

3) Знайти середню вагу дітей для усіх можливих поєднань параметрів
smoker та mage35.

In [0]:
df3 = df \
    .groupBy("smoker", "mage35") \
    .agg(F.avg("Birthweight").alias('AverageWeight'))
display(df3)

smoker mage35 AverageWeight 0 1 4.1 1 1 2.783333333333333 0 0 3.478421052631579 1 0 3.189473684210526

4) Додати в таблицю Birthweight новий стовпець, що містить різницю
між середнім та поточним значенням ваги дитини в залежності від
того чи палить батько. Наприклад, якщо середня вага дітей, у яких
батько не палить, 3.55, а в першому рядку вага 3.5 і батько не палить,
то значення в додатковому стовпчику = 3.55-3.5=0.05. Зберегти
новий файл.

In [0]:
df4 = df.withColumn("fsmoker", F.when((F.col("fnocig") > 0), 1).otherwise(0))
display(df4)

ID Length Birthweight Headcirc Gestation smoker mage mnocig mheight mppwt fage fedyrs fnocig fheight lowbwt mage35 fsmoker 1360 56 4.55 34 44 0 20 0 162 57 23 10 35 179 0 0 1 1016 53 4.32 36 40 0 19 0 171 62 19 12 0 183 0 0 0 462 58 4.1 39 41 0 35 0 172 58 31 16 25 185 0 1 1 1187 53 4.07 38 44 0 20 0 174 68 26 14 25 189 0 0 1 553 54 3.94 37 42 0 24 0 175 66 30 12 0 184 0 0 0 1636 51 3.93 38 38 0 29 0 165 61 31 16 0 180 0 0 0 820 52 3.77 34 40 0 24 0 157 50 31 16 0 173 0 0 0 1191 53 3.65 33 42 0 21 0 165 61 21 10 25 185 0 0 1 1081 54 3.63 38 38 0 18 0 172 50 20 12 7 172 0 0 1 822 50 3.42 35 38 0 20 0 157 48 22 14 0 179 0 0 0 1683 53 3.35 33 41 0 27 0 164 62 37 14 0 170 0 0 0 1088 51 3.27 36 40 0 24 0 168 53 29 16 0 181 0 0 0 1107 52 3.23 36 38 0 31 0 164 57 35 16 0 183 0 0 0 755 53 3.2 33 41 0 21 0 155 55 25 14 25 183 0 0 1 1058 53 3.15 34 40 0 29 0 167 60 30 16 25 182 0 0 1 321 48 3.11 33 37 0 28 0 158 54 39 10 0 171 0 0 0 697 48 3.03 35 39 0 27 0 162 62 27 14 0 178 0 0 0 808 48 2.92 33 34 0 26 0 167 64 25 12 25 175 0 0 1 1600 53 2.9 34 39 0 19 0 165 57 23 14 2 193 0 0 1 1313 43 2.65 32 33 0 24 0 149 45 26 16 0 169 1 0 0 792 53 3.64 38 40 1 20 2 170 59 24 12 12 185 0 0 1 1388 51 3.14 33 41 1 22 7 160 53 24 16 12 176 0 0 1 575 50 2.78 30 37 1 19 7 165 60 20 14 0 183 0 0 0 569 50 2.51 35 39 1 22 7 159 52 23 14 25 200 1 0 1 1363 48 2.37 30 37 1 20 7 163 47 20 10 35 185 1 0 1 300 46 2.05 32 35 1 41 7 166 57 37 14 25 173 1 1 1 431 48 1.92 30 33 1 20 7 161 50 20 10 35 180 1 0 1 1764 58 4.57 39 41 1 32 12 173 70 38 14 25 180 0 0 1 532 53 3.59 34 40 1 31 12 163 49 41 12 50 191 0 0 1 752 49 3.32 36 40 1 27 12 152 48 37 12 25 170 0 0 1 1023 52 3 35 38 1 30 12 165 64 38 14 50 180 0 0 1 57 51 3.32 38 39 1 23 17 157 48 32 12 25 169 0 0 1 1522 50 2.74 33 39 1 21 17 156 53 24 12 7 179 0 0 1 223 50 3.87 33 45 1 28 25 163 54 30 16 0 183 0 0 0 272 52 3.86 36 39 1 30 25 170 78 40 16 50 178 0 0 1 27 53 3.55 37 41 1 37 25 161 66 46 16 0 175 0 1 0 365 52 3.53 37 40 1 26 25 170 62 30 10 25 181 0 0 1 619 52 3.41 33 39 1 23 25 181 69 23 16 2 181 0 0 1 1369 49 3.18 34 38 1 31 25 162 57 32 16 50 194 0 0 1 1262 53 3.19 34 41 1 27 35 163 51 31 16 25 185 0 0 1 516 47 2.66 33 35 1 20 35 170 57 23 12 50 186 1 0 1 1272 53 2.75 32 40 1 37 50 168 61 31 16 0 173 0 1 0

In [0]:
df4_tmp = df4 \
    .groupby("fsmoker") \
    .agg(F.avg("BirthWeight").alias('AverageWeight')) \
    .withColumnRenamed("fsmoker", "indicator")

display(df4_tmp)

indicator AverageWeight 1 3.265555555555555 0 3.398

In [0]:
df4 = df4 \
    .join(df4_tmp, df4.fsmoker ==  df4_tmp.indicator, "inner") \
    .withColumn("result",F.col("AverageWeight") - F.col("Birthweight")) \
    .select("ID", "Birthweight", "AverageWeight", "fsmoker", "result")
display(df4)

ID Birthweight AverageWeight fsmoker result 1360 4.55 3.265555555555555 1 -1.284444444444445 1016 4.32 3.398 0 -0.9220000000000002 462 4.1 3.265555555555555 1 -0.8344444444444448 1187 4.07 3.265555555555555 1 -0.8044444444444454 553 3.94 3.398 0 -0.5419999999999998 1636 3.93 3.398 0 -0.532 820 3.77 3.398 0 -0.3719999999999999 1191 3.65 3.265555555555555 1 -0.38444444444444503 1081 3.63 3.265555555555555 1 -0.364444444444445 822 3.42 3.398 0 -0.021999999999999797 1683 3.35 3.398 0 0.04800000000000004 1088 3.27 3.398 0 0.1280000000000001 1107 3.23 3.398 0 0.16800000000000015 755 3.2 3.265555555555555 1 0.0655555555555547 1058 3.15 3.265555555555555 1 0.11555555555555497 321 3.11 3.398 0 0.28800000000000026 697 3.03 3.398 0 0.3680000000000003 808 2.92 3.265555555555555 1 0.34555555555555495 1600 2.9 3.265555555555555 1 0.36555555555555497 1313 2.65 3.398 0 0.7480000000000002 792 3.64 3.265555555555555 1 -0.37444444444444525 1388 3.14 3.265555555555555 1 0.12555555555555475 575 2.78 3.398 0 0.6180000000000003 569 2.51 3.265555555555555 1 0.7555555555555551 1363 2.37 3.265555555555555 1 0.8955555555555548 300 2.05 3.265555555555555 1 1.215555555555555 431 1.92 3.265555555555555 1 1.345555555555555 1764 4.57 3.265555555555555 1 -1.3044444444444454 532 3.59 3.265555555555555 1 -0.324444444444445 752 3.32 3.265555555555555 1 -0.05444444444444496 1023 3 3.265555555555555 1 0.2655555555555549 57 3.32 3.265555555555555 1 -0.05444444444444496 1522 2.74 3.265555555555555 1 0.5255555555555547 223 3.87 3.398 0 -0.472 272 3.86 3.265555555555555 1 -0.594444444444445 27 3.55 3.398 0 -0.1519999999999997 365 3.53 3.265555555555555 1 -0.2644444444444449 619 3.41 3.265555555555555 1 -0.14444444444444526 1369 3.18 3.265555555555555 1 0.08555555555555472 1262 3.19 3.265555555555555 1 0.07555555555555493 516 2.66 3.265555555555555 1 0.6055555555555547 1272 2.75 3.398 0 0.6480000000000001

In [0]:
df4.write.csv("/FileStore/tables/result.csv")